icicle/backend/cpu/include/ntt_tasks.h

```diff
@@ -0,0 +1,1260 @@
+   */
+
+  template <typename S, typename E>
+  void NttCpu<S, E>::refactor_output_h0(E* elements, NttTaskCordinates ntt_task_cordinates)
+  {
+    int h0_subntt_size =
+      1 << NttCpu<S, E>::ntt_sub_logn.hierarchy_0_layers_sub_logn[ntt_task_cordinates.hierarchy_1_layer_idx]
+                                                                 [ntt_task_cordinates.hierarchy_0_layer_idx];
+    int h0_nof_subntts =
+      1 << NttCpu<S, E>::ntt_sub_logn
+             .hierarchy_0_layers_sub_logn[ntt_task_cordinates.hierarchy_1_layer_idx][0]; // only relevant for layer 1
+    int i, j, i_0;
+    int ntt_size =
+      ntt_task_cordinates.hierarchy_0_layer_idx == 0
+        ? 1
+            << (NttCpu<S, E>::ntt_sub_logn.hierarchy_0_layers_sub_logn[ntt_task_cordinates.hierarchy_1_layer_idx][0] +
+                NttCpu<S, E>::ntt_sub_logn.hierarchy_0_layers_sub_logn[ntt_task_cordinates.hierarchy_1_layer_idx][1])
+        : 1
+            << (NttCpu<S, E>::ntt_sub_logn.hierarchy_0_layers_sub_logn[ntt_task_cordinates.hierarchy_1_layer_idx][0] +
+                NttCpu<S, E>::ntt_sub_logn.hierarchy_0_layers_sub_logn[ntt_task_cordinates.hierarchy_1_layer_idx][1] +
+                NttCpu<S, E>::ntt_sub_logn.hierarchy_0_layers_sub_logn[ntt_task_cordinates.hierarchy_1_layer_idx][2]);
+    int stride = this->config.columns_batch ? this->config.batch_size : 1;
+    uint64_t original_size = (1 << NttCpu<S, E>::ntt_sub_logn.logn);
+    for (int batch = 0; batch < this->config.batch_size; ++batch) {
+      E* h1_subntt_elements =
+        elements +
+        stride * (ntt_task_cordinates.hierarchy_1_subntt_idx
+                  << NttCpu<S, E>::ntt_sub_logn
+                       .hierarchy_1_layers_sub_logn[ntt_task_cordinates.hierarchy_1_layer_idx]); // input + subntt_idx *
+                                                                                                 // subntt_size
+      E* elements_of_current_batch =
+        this->config.columns_batch ? h1_subntt_elements + batch : h1_subntt_elements + batch * original_size;
+      for (int elem = 0; elem < h0_subntt_size; elem++) {
+        uint64_t elem_mem_idx = stride * idx_in_mem(ntt_task_cordinates, elem);
+        i = (ntt_task_cordinates.hierarchy_0_layer_idx == 0)
+              ? elem
+              : elem * h0_nof_subntts + ntt_task_cordinates.hierarchy_0_subntt_idx;
+        j = (ntt_task_cordinates.hierarchy_0_layer_idx == 0) ? ntt_task_cordinates.hierarchy_0_subntt_idx
+                                                             : ntt_task_cordinates.hierarchy_0_block_idx;
+        uint64_t tw_idx = (this->direction == NTTDir::kForward)
+                            ? ((this->domain_max_size / ntt_size) * j * i)
+                            : this->domain_max_size - ((this->domain_max_size / ntt_size) * j * i);
+        elements_of_current_batch[elem_mem_idx] = elements_of_current_batch[elem_mem_idx] * this->twiddles[tw_idx];
```

Comment:
have special memory for twiddles for cash efficiency instead of reading from global